### PR TITLE
Features/device controller

### PR DESCRIPTION
**Pull request by kamilkusmierz**
Add device,hardware,crop packages

### CODE REVIEW

**Comment by rksiazek** [comment]
Something is wrong with these tests...
I think we have to review and refactor this code thoroughly :man_factory_worker:

File: src/test/java/com/iotg/greenhouse/backendapp/Auth/AuthControllerRestTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@AutoConfigureMockMvc
+class AuthControllerRestTest {
```
Comment (rksiazek) [inline src/test/java/com/iotg/greenhouse/backendapp/Auth/AuthControllerRestTest.java]: These are integration tests and should be separated from unit tests.

File: src/test/java/com/iotg/greenhouse/backendapp/Auth/AuthJPAServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    @InjectMocks
+    private AuthJPAService authJPAService;
+    @Mock
+    private UserJPARepository userJPARepository;
+    @Mock
+    private RoleJPARepository roleJPARepository;
+
+
+    @Test
+    void existsByUsernameTest() {
```
Comment (rksiazek) [inline src/test/java/com/iotg/greenhouse/backendapp/Auth/AuthJPAServiceTest.java]: Test methods' names should be standardized throughout entire scope of the project
( fe. givenThatUserExists_whenCheckByNameIfExists_thenReturnTrue ).

**Comment by rksiazek** [comment]
Could you squash this commit https://github.com/iot-greenhouse/iot-greenhouse-backend/pull/13/commits/51a59468b2bc74604fcd05040fb1a2ee685b05d7 with the precending one?

**Comment by rksiazek** [comment]
We will take care of code styling standards in another PR as it could mess this one.

**Comment by rksiazek** [comment]
[[ Backend ] Szkielet aplikacji](https://trello.com/c/A4VWKF1r/8-backend-szkielet-aplikacji)

File: src/main/java/com/iotg/greenhouse/backendapp/models/ResponseMessage.java [hunk 3]
```diff
@@ -1,26 +1,27 @@
 package com.iotg.greenhouse.backendapp.models;
```
Comment (rksiazek) [inline src/main/java/com/iotg/greenhouse/backendapp/models/ResponseMessage.java]: I see that this class is not used anywhere and it doesn't seem to be necessary for this project. Maybe we should delete it?

File: src/main/java/com/iotg/greenhouse/backendapp/Device/DeviceJPARepository.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Repository
+public interface DeviceJPARepository extends JpaRepository<Device, Long> {
+
+    <T> List<T> findDevicesByGreenhouse_Id(Long id, Class<T> type);
+
+    <T> T findDeviceByDeviceIdAndGreenhouse_Id(String deviceId, Long id, Class<T> type);
+
+    <T> T findESPDeviceByDeviceIdAndGreenhouse_Name(String deviceId, String insId, Class<T> type);
+
+    <T> T findDeviceById(Long id, Class<T> type);
```
[the remark this answers is not in the document]
Comment (kamilkusmierz) [reply src/main/java/com/iotg/greenhouse/backendapp/Device/DeviceJPARepository.java]: Methods are correct.

File: src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerImp.java [hunk 5]
```diff
@@ -1,42 +1,58 @@
...
     @Override
     public List<GreenhouseDTO> fetchAllGreenhouse() {
         return greenhouseJPAService.fetchAllGreenHousesByUser();
     }
 
     @Override
-    public GreenhouseDTO updateGreenhouse(GreenHouse greenhouse) throws Exception {
+    public GreenhouseDTO updateGreenhouse(Greenhouse greenhouse) throws Exception {
         Long id = greenhouse.getId();
         GreenhouseDTO insByName = greenhouseJPAService.fetchGreenHouseByName(greenhouse.getName());
```
[the remark this answers is not in the document]
Comment (kamilkusmierz) [reply src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerImp.java]: It should be greenhouseFoundByName

File: src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerREST.java [hunk 6]
```diff
@@ -1,51 +1,52 @@
...
+    public ResponseEntity<?> getConfig(@PathVariable("greenhouseId") Long greenhouseId, @PathVariable("deviceId") Long deviceId) {
+        return new ResponseEntity<>(greenhouseController.getConfig(greenhouseId, deviceId), HttpStatus.OK);
+    }
 
     @PreAuthorize("hasRole('USER')")
-    @GetMapping("/fetchAllGreenhouse")
+    @GetMapping("/Greenhouse")
     public ResponseEntity<?> fetchAllGreenHouse() {
-        return new ResponseEntity<List<GreenhouseDTO>>(greenhouseController.fetchAllGreenhouse(), HttpStatus.OK);
+        return new ResponseEntity<>(greenhouseController.fetchAllGreenhouse(), HttpStatus.OK);
```
[the remark this answers is not in the document]
Comment (kamilkusmierz) [reply src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerREST.java]: to make it flexibility

File: src/main/java/com/iotg/greenhouse/backendapp/models/ResponseMessage.java (hunk 3 above)
Comment (kamilkusmierz) [reply src/main/java/com/iotg/greenhouse/backendapp/models/ResponseMessage.java]: yes we should

File: src/main/java/com/iotg/greenhouse/backendapp/Device/DeviceControllerImp.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        Optional<Greenhouse> greenhouse = greenhouseJPAService.fetchGreenHouseById(greenhouseId);
+        if (greenhouse.isPresent()) {
+            espDevice.setGreenhouse(greenhouseJPAService.fetchGreenHouseById(greenhouseId)
+                    .get());
+        } else {
+            throw new Exception("Instance not found");
+        }
+
+
+        Device foundByDeviceId = deviceJPAService.fetchDeviceByDeviceIdAndGreenhouseId(deviceId, greenhouseId, Device.class);
```
[the remark this answers is not in the document]
Comment (rksiazek) [reply src/main/java/com/iotg/greenhouse/backendapp/Device/DeviceControllerImp.java]: Correction: code styling related issues will be fixed in another PR.

File: src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerImp.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (rksiazek) [reply src/main/java/com/iotg/greenhouse/backendapp/Greenhouse/GreenhouseControllerImp.java]: Lets call it simply `greenhouse`.

**Pull request merged**